nixos/modules/tasks/filesystems/zfs.nix

```diff
@@ -391,9 +414,24 @@ in
               Type = "oneshot";
               RemainAfterExit = true;
             };
-            script = ''
-              zpool_cmd="${packages.zfsUser}/sbin/zpool"
-              ("$zpool_cmd" list "${pool}" >/dev/null) || "$zpool_cmd" import -d ${cfgZfs.devNodes} -N ${optionalString cfgZfs.forceImportAll "-f"} "${pool}"
+            script = (importLib {
+              zpoolCmd="${packages.zfsUser}/sbin/zpool";
+              awkCmd="${pkgs.gawk}/bin/awk";
+              inherit pool cfgZfs;
+            }) + ''
+              imported && exit
+              echo -n "importing ZFS pool \"${pool}\"..."
+              # Loop across the import until it succeeds, because the devices needed may not be discovered yet.
+              for trial in `seq 1 60`; do
+                ready && import && break
+                sleep 1
+              done
+              imported || import  # Try one last time, e.g. to import a degraded pool.
+              if imported; then
```

Comment:
What is the return value of `zpool import` in case of an failure? Should the script not already exit before it reaches this line?

Reply:
Most likely. I'm going to call this defensive coding; no matter what happens, we won't claim the pool is successfully imported if it isn't.